prime-router/docs/proposals/0005-api-hardening.md

```diff
@@ -0,0 +1,125 @@
+# API Endpoint Hardening Proposal
+
+## Context and Motivation
+
+As a secure system, the Prime Data Hub's API end-points must be hardened against malicious attacks that manipulate the payload sent through the data hub. 
+The API end-points should have a number of limits, checks and validations that prevents from an attacker who knows about the inner workings of the data hub (ie. read our source code) from information disclosure and service disruption. 
+
+This epic is not about DDOS attacks which will be handled by the WAF epics. 
+
+## Goals 
+
+1. Enforce limits on report size overall size, and on individual data elements size
+2. Prevent SQL and Javascript injection, by design.
+
+## Detailed proposal for overall size limitation, and individual field size limitations
+
+The system will define these overall limits:
+
+- MAX_BYTES - a reasonable limit on the overall size of the payload. (Azure limit is 100Meg.)   **Suggest: 40Meg. (allows 4k per row)** 
+- MAX_ITEMS - a limit on the number of Items (in a CSV, this is a limit on the number of rows + 1).  **Suggest: 10,000 Items (10,001 rows)**
+- **Phase 2 work?**: (MAX_COLUMNS - for CSVs only, a limit on the number of columns input. **Suggest: 1000 columns**)
+- MAX_ITEM_SIZE - byte limit on the size of one item.  This is most useful for reading in hl7 - we'll want a way to reject inputs without having to get bogged down in detailed item parsing.   **Suggest:  200K, to match Redox max**
+- MAX_ERRORS - Validation will fail immediately if the total number of Errors passes this threshold.  **Suggest: 100 errors**
+- Note:  no limit on warnings, since warnings by definition should not cause failures.
+
+These limits will be set globally, hardcoded, for now.  Could be set per schema in the future, with careful consideration of how that gets locked down.
+
+The limits will apply to both CSV and HL7, when appropriate.
```

Comment:
In the case if HL7, instead of speaking of 10,000 items/10,001 rows, we will have to limit it to some number of messages, and it should be less than 10,000. From my testing, that many faked HL7 messages is just over 40MB.

Reply:
This is very helpful.   Very roughly, what would be the total Megs of 10,000 patients worth of test results, in hl7?  Maybe that should be our `MAX_BYTES`.   I realize you can only estimate, but we could round up from a ballpark number.